File: frontend/src/utils/text.js
/**
 * QUAEST.TECH — text sanitizers
 * The backend still emits some emoji-prefixed strings (reason text, catalyst
 * labels, direction labels). Until those are stripped at generation
 * (backend follow-up: scanner.py / routers/market.py), we strip them on
 * display so the UI honors the "zero emoji" rule.
 */

// Broad emoji + symbol ranges (pictographs, dingbats, arrows, flags, VS16).
const EMOJI_RE =
  /[\u{1F000}-\u{1FAFF}\u{2600}-\u{27BF}\u{2B00}-\u{2BFF}\u{2190}-\u{21FF}\u{2300}-\u{23FF}\u{25A0}-\u{25FF}\u{1F1E6}-\u{1F1FF}\u{FE0F}\u{200D}]/gu;

/** Remove every emoji/pictograph from a string, collapse the leftover space. */
export function stripEmoji(s) {
  if (typeof s !== 'string') return s;
  return s.replace(EMOJI_RE, '').replace(/\s{2,}/g, ' ').trim();
}

File: api/core/text.py
"""
Text sanitizers — keep the API emoji-free at the source.

The UI mandate is "zero emoji" (UI_OVERHAUL_PLAN §1). Rather than strip emoji
client-side (the old `frontend/src/utils/text.stripEmoji` shim), we strip them
where the strings are *generated* — score reasons, signals, catalyst/direction
labels — so the API never emits a pictograph in the first place.

`strip_emoji` is also applied defensively in `db.models.save_scan` when signal
text is persisted, catching any straggler from `intel/` that bypassed the
generation-site cleanup.
"""

import re

# Pictographs, dingbats, arrows, geometric shapes, flags, variation selector,
# zero-width joiner. Mirrors the frontend EMOJI_RE so behaviour is identical on
# both sides during the migration (frontend shim becomes dead code afterwards).
_EMOJI_RE = re.compile(
    "["
    "\U0001F000-\U0001FAFF"   # symbols & pictographs, supplemental, extended-A
    "\U00002600-\U000027BF"   # misc symbols + dingbats
    "\U00002B00-\U00002BFF"   # misc symbols and arrows
    "\U00002190-\U000021FF"   # arrows
    "\U00002300-\U000023FF"   # technical (⚡ ⟨⟩ etc.)
    "\U000025A0-\U000025FF"   # geometric shapes
    "\U0001F1E6-\U0001F1FF"   # regional indicators (flags)
    "\U0000FE0F"              # variation selector-16
    "\U0000200D"              # zero-width joiner
    "]",
    flags=re.UNICODE,
)

_MULTISPACE_RE = re.compile(r"\s{2,}")


def strip_emoji(s):
    """Remove every emoji/pictograph from a string and collapse leftover spaces.

    Non-string input is returned unchanged so this is safe to map over mixed
    values without guarding each call site.
    """
    if not isinstance(s, str):
        return s
    return _MULTISPACE_RE.sub(" ", _EMOJI_RE.sub("", s)).strip()
